Add support for `max_version, dl_device, copy` kwargs in `__dlpack__` to match Array API

Micky774 opened the pull request:
Towards https://github.com/google/jax/issues/20200

cf. https://github.com/data-apis/array-api/pull/741, https://github.com/data-apis/array-api/pull/602, https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html

> [!NOTE]
> "In principle, arbitrary cross-device copies could be allowed too, but the consensus in https://github.com/data-apis/array-api/issues/626 was that limiting to device-to-host copies is enough for now". This PR _includes_ the optional device-to-device transfer.

Default behavior is preserved when `max_version=None, dl_device=None, copy=None, `

This PR also adds new versioning information to `jax/_src/dlpack.py`:
```python
DLPACK_VERSION = (0, 8)
MIN_DLPACK_VERSION = (0, 5)
```

yashk2810 commented on jax/_src/array.py:
```diff
@@ -404,11 +404,54 @@ def __array__(self, dtype=None, context=None, copy=None):
     from jax._src.dlpack import to_dlpack  # pylint: disable=g-import-not-at-top
-    return to_dlpack(self, stream=stream)
+
+    device_set = self.sharding.device_set
```
Since you are already importing things from dlpack here, can this logic live in that file and you just call that function here?

jakevdp commented on jax/_src/array.py:
```diff
@@ -404,11 +404,54 @@ def __array__(self, dtype=None, context=None, copy=None):
+      else:
+        arr = api.device_put(self, dlpack_device)
+    if arr is None:
+        arr = self.copy() if copy else self
```
two-space indent

jakevdp commented on jax/_src/dlpack.py:
```diff
@@ -78,9 +85,17 @@ def to_dlpack(x: Array, take_ownership: bool = False,
-  )  # type: ignore
+
+  if max_version is None or max_version[0] >= DLPACK_VERSION[0]:
+      return _to_dlpack(x, stream)
```
2-space indent

jakevdp commented on jax/_src/dlpack.py:
```diff
@@ -78,9 +85,17 @@ def to_dlpack(x: Array, take_ownership: bool = False,
+      return _to_dlpack(x, stream)
+  elif max_version >= MIN_DLPACK_VERSION:
+    # Legacy path
+    pass
```
It seems strange to return `None` in this case

Micky774 replied:
This branch is never hit and is really just a template for if/when we add a legacy path

jakevdp replied:
What is the legacy path? Is this something that users could conceivably hit? If so then we should do something like this:
```python
raise NotImplementedError("dlpack legacy path is not implemented")
```
If this branch should truly never be hit by any conceivable user code, then I'd do something like
```python
raise RuntimeError("This branch should be unreachable. Please open a bug if you see this.")
```

Micky774 replied:
Right now I've set both `DLPACK_VERSION` and `MIN_DLPACK_VERSION` to `(0, 1)`, so currently the path is inaccessible. When we add a way to produce the newer `DLManagedTensorVersioned` format in XLA, then we increment `DLPACK_VERSION` and this path will actually be usable and the _current_ methodology will become the legacy path -- i.e. `_to_dlpack() --> _legacy_to_dlpack()` and we we'll have a new `_to_dlpack()` to match the altered XLA DLPack producer.

In the interim I will add the `RuntimeError` you suggested so that its role is explicit :)

jakevdp commented on jax/_src/array.py:
```diff
@@ -404,11 +404,27 @@ def __array__(self, dtype=None, context=None, copy=None):
+                 dl_device: tuple[enum.Enum, int] | None = None,
+                 copy: bool | None = None):
     if len(self._arrays) != 1:
       raise BufferError("__dlpack__ only supported for unsharded arrays.")
```
I think this check is better covered by the `device_set` check below. We could remove this.

jakevdp commented on jax/_src/dlpack.py:
```diff
@@ -48,9 +51,33 @@ class DLDeviceType(enum.IntEnum):
+      )
+    else:
+      arr = device_put(x, dlpack_device)
+  if arr is None:
```
Would `else:` suffice here? As a bonus, then `arr` could be typed `Array` and the `# type: ignore` below may not be necessary.

jakevdp commented on jax/_src/dlpack.py:
```diff
@@ -73,14 +100,40 @@ def to_dlpack(x: Array, take_ownership: bool = False,
+      }[dl_device_type]
+      backend = xla_bridge.get_backend(dl_device_platform)
+      dlpack_device = backend.device_from_local_hardware_id(local_hardware_id)
+    except TypeError:
```
Do `except TypeError as err` and below `raise BufferError(...) from err`.

yashk2810 commented on jax/_src/array.py:
```diff
@@ -404,11 +404,27 @@ def __array__(self, dtype=None, context=None, copy=None):
+        f"device, but an array with a Sharding over {len(device_set)} devices "
+        "was provided."
+      )
+    device = device_set.pop()
```
Do `device, = device_set` instead of pop?

jakevdp commented:
Minor comment, but in general it's useful to have meaningful commit messages ("Update" doesn't communicate much about what the change includes).

Micky774 commented:
@yashk2810 @jakevdp This PR should be ready for review again. I've added tests and updated the documentation of `to_dlpack`. Note that I have purposefully left the `max_version` resolving logic expanded and verbose despite only having one path so that it is easy and obvious to update as DLPack moves to version 1.0 in the near future (and more importantly, as XLA eventually supports it). If you'd prefer I just collapse it down for now and leave it for future PRs, let me know and I'll update 😄

jakevdp commented:
Tests are failing because `jax2tf` uses the `take_ownership` parameter removed in this PR: https://github.com/google/jax/blob/abfbb0ae2b85a6c887caa5d11609b347ee6b8f2d/jax/experimental/jax2tf/call_tf.py#L337

Micky774 committed:
Add support for max_version, dl_device, copy kwargs in __dlpack__

Micky774 commented:
@jakevdp Done -- I've also simplified the tests a bit and refactored `DLDeviceType` to `jax._src.typing`
